src/System.Memory/ref/System.Memory.cs

```diff
@@ -184,6 +184,52 @@ public partial struct Enumerator
+    {
+        public SequenceReader(System.Buffers.ReadOnlySequence<T> buffer) { }
+        public bool IsLastSegment { get { throw null; } }
+        public bool End { get { throw null; } }
+        public ReadOnlySequence<T> Sequence { get { throw null; } }
+        public SequencePosition Position { get { throw null; } }
+        public ReadOnlySpan<T> CurrentSpan { get { throw null; } }
+        public int CurrentSpanIndex { get { throw null; } }
+        public ReadOnlySpan<T> UnreadSpan { get { throw null; } }
+        public long Consumed { get { throw null; } }
+        public long Remaining { get { throw null; } }
+        public long Length { get { throw null; } }
+        public ReadOnlySpan<T> Peek(Span<T> copyBuffer) { throw null; }
+        public bool TryPeek(out T value) { throw null; }
+        public void Advance(long count) { throw null; }
+        public void Rewind(long count) { throw null; }
+        public bool TryRead(out T value) { throw null; }
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter) { throw null; }
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter) { throw null; }
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, T delimiterEscape, bool advancePastDelimiter = true) { throw null; }
+        public bool TryReadTo(out ReadOnlySequence<T> sequence, T delimiter, bool advancePastDelimiter = true) { throw null; }
+        public bool TryReadTo(out ReadOnlySequence<T> sequence, T delimiter, T delimiterEscape, bool advancePastDelimiter = true) { throw null; }
+        public bool TryReadToAny(out ReadOnlySpan<T> span, ReadOnlySpan<T> delimiters, bool advancePastDelimiter = true) { throw null; }
+        public bool TryReadToAny(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> delimiters, bool advancePastDelimiter = true) { throw null; }
+        public bool TryReadTo(out ReadOnlySequence<T> sequence, ReadOnlySpan<T> delimiter, bool advancePastDelimiter = true) { throw null; }
+        public bool TryAdvanceTo(T delimiter, bool advancePastDelimiter = true) { throw null; }
+        public bool TryAdvanceToAny(ReadOnlySpan<T> delimiters, bool advancePastDelimiter = true) { throw null; }
+        public long AdvancePast(T value) { throw null; }
+        public long AdvancePastAny(T value0, T value1) { throw null; }
+        public long AdvancePastAny(T value0, T value1, T value2) { throw null; }
+        public long AdvancePastAny(T value0, T value1, T value2, T value3) { throw null; }
+        public long AdvancePastAny(ReadOnlySpan<T> values) { throw null; }
+        public bool IsNext(T next, bool advancePast = false) { throw null; }
+        public bool IsNext(ReadOnlySpan<T> next, bool advancePast = false) { throw null; }
+    }
+    public static partial class SequenceReaderExtensions
+    {
+        public static bool TryRead<T>(ref this SequenceReader<byte> reader, out T value) where T : unmanaged { throw null; }
```

Comment:
The equivalent functionality for `Span<byte>` is on the `MemoryMarshal` class, and IIRC we put it there for safety concerns.  Are those concerns not relevant here?  And are there no endianness issues here? I remember we had long conversations about this one, and it ended up moving late in the 2.1 cycle.
cc: @ahsonkhan, @KrzysztofCwalina, @jkotas

Reply:
The explicit endian methods are there. When you're reading structs out of a data stream this is particularly useful.

Reply:
And particularly dangerous.

Reply:
> And particularly dangerous.

For what exactly? I get that you can get endianness wrong.

Reply:
Discussed in https://github.com/dotnet/corefx/issues/27094

The `unmanaged` constraint is not particularly useful for API like this. It is too dangerous and too limiting at the same time.

Reply:
So we're just talking about reading in state to some struct that expects the data to be within a certain set of constraints? Or is there more to it? The API is super useful for parsing data per some specification- I'm fine with rubbing the danger in people's faces via naming / copious comments.

Reply:
Right, we're talking about reading in state to some struct that expects the data to be within a certain set of constraints. It the data are coming from untrusted source (the common case), the violation of these constrains can lead to buffer overruns and other types of security bugs.

Reply:
Added explicit comments. We can iterate on naming of the method as well if we feel it necessary.

Reply:
If we do change the name here we should consider adding additional overloads for intrinsic types we would expect to find in binary format (integers, etc.). Not wanting a bunch of overloads was what kept me from adding this as Interop etc.

Reply:
> If we do change the name here we should consider adding additional overloads for intrinsic types we would expect to find in binary format (integers, etc.)

Most of the overloads are there already. It would be best if none of the methods on SequenceReader used current machine endianess.

Reply:
> It would be best if none of the methods on SequenceReader used current machine endianess.

Ok.

Reply:
Moving this API to SequenceMarshal.